### ISSUE TITLE
Stick-table / CLI: permit to remove entries with complicated keys

### ISSUE THREAD

**Issue by sawall** [haproxy/haproxy#2118]
### Detailed Description of the Problem

Somehow a blank entry appears to have ended up in my stick table:

```
$ echo "show table nodes" | sudo -u haproxy socat stdio /var/run/haproxy/admin.sock
# table: nodes, type: string, size:204800, used:1556
0x55b863c31250: key= use=0 exp=299997 server_id=38 server_key=10.XX.XX.XX:443
...
```

I can't delete it using the runtime API using "clear table" because it won't take blank. This is causing all traffic that isn't stuck elsewhere in a subdirectory to all forward to the same server.

This is partly a bug report since I believe the stick table should never have a blank entry but also a request for help for how to eliminate this from my stick table since this mesh is fronting for tens of thousands of production connections and we would really prefer to not have to take the entire ingress offline to resolve this.

This entry has been in the stick tables for several months now, I first noticed it and tried to get help on the forum in December: https://discourse.haproxy.org/t/stick-table-has-empty-key/8188

### Expected Behavior

I expect blank entries to not stick.

### Steps to Reproduce the Behavior

I have no idea what caused this.

### What is your configuration?

The relevant backend configurations include the following. In the frontend X-StickToken is set as is X-StickTokenFlag. In some cases X-StickToken and X-StickTokenFlag may both come in with blank strings.

```
backend nodes
  acl stick_token_flag hdr_reg(X-StickTokenFlag) .+
  option forwardfor

  balance hdr(X-StickToken)
  hash-type consistent djb2
  hash-balance-factor 150

  stick-table type string len 200 size 200k peers mypeers expire 5m srvkey addr
  stick match hdr(X-StickToken) table nodes
  stick on hdr(X-StickToken) table nodes if stick_token_flag
  
  ....

backend release-3958
  acl stick_token_flag hdr_reg(X-StickTokenFlag) .+
  option forwardfor      # pass client IP to server
  no option allbackups   # use only the first backup server to ensure fallbacks will be by release
  default-server maxconn 2000000 inter 2s fall 3 rise 10

  # custom http health checks for our nodes
  option httpchk
  http-check send meth GET uri /
  http-check expect status 200

  balance hdr(X-StickToken)
  hash-type consistent djb2
  hash-balance-factor 150
  stick match hdr(X-StickToken) table nodes
  stick on hdr(X-StickToken) table nodes if stick_token_flag
  ```

### Output of `haproxy -vv`

```plain
$ haproxy -vv
HAProxy version 2.6.9-1ppa1~jammy 2023/02/16 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2027.
Known bugs: http://www.haproxy.org/bugs/bugs-2.6.9.html
Running on: Linux 5.15.0-1029-oracle #35-Ubuntu SMP Tue Jan 24 15:17:52 UTC 2023 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -O2 -flto=auto -ffat-lto-objects -flto=auto -ffat-lto-objects -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_SYSTEMD=1 USE_OT=1 USE_PROMEX=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : -51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL -ENGINE +EPOLL -EVPORTS +GETADDRINFO -KQUEUE +LIBCRYPT +LINUX_SPLICE +LINUX_TPROXY +LUA -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL +OT -PCRE +PCRE2 +PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PROCCTL +PROMEX -QUIC +RT +SLZ -STATIC_PCRE -STATIC_PCRE2 +SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL -ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=4).
Built with OpenSSL version : OpenSSL 3.0.2 15 Mar 2022
Running on OpenSSL version : OpenSSL 3.0.2 15 Mar 2022
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
OpenSSL providers loaded : default
Built with Lua version : Lua 5.3.6
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with OpenTracing support.
Support for malloc_trim() is enabled.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.39 2021-10-29
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 11.3.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=

Available services : prometheus-exporter
Available filters :
	[CACHE] cache
	[COMP] compression
	[FCGI] fcgi-app
	[  OT] opentracing
	[SPOE] spoe
	[TRACE] trace
```

**Comment by wtarreau** [comment]
I indeed can't figure a quick solution to clear such an entry due to the limitations of the CLI language. Maybe it can work to only remove entries sticking to this server with "clear table nodes data.server_id eq 38" but I'm not certain.

What is causing this clearly is that you're getting a header field with this empty value. Maybe you'd need a rule to ignore it when it's empty. The right way to proceed would be to update the config and reload. The table will be transferred to the new process via the peers protocol and since you won't match anymore an empty header the entry won't be used anymore and will vanish. Note that your `balance hdr()` on it may also cause some imbalance if the header is sometimes empty.

I'm seeing one thing which is not correct in your config in fact. `stick on` is exactly one `stick match` and one `stick store`. So here you seem to have made this check for non-emptiness on the `stick on` rule, but since you didn't put it on the `stick match` the entry is regularly looked up and refreshed. So most likely you just need to append `if stick_token_flag` to your `stick match` rule (and probably rename your `stick on` to `stick store`).

**Comment by sawall** [comment]
Thanks so much for taking a look at my configuration. I'll put a conditional check on `stick match` which will hopefully clean this up.

Deleting all entries for the server is a little better than dumping the entire stick table but isn't ideal. I also now realize after talking this through that our real problem is the misconfiguration that leads to the stick table entry never expiring. In one test environment this has persisted so long that it appears to have led to a bad entry that's pointing at a now-nonexistent server.

Unless there is an intent to implement a way to interact with stick table entries based on id instead of by key, I think I can just close this since it's resolvable via a config fix.

**Comment by wtarreau** [comment]
Indeed I think we can turn this one into a feature request to permit removal of such entries via the CLI because I definitely agree that it's a pain as-is.

**Comment by wtarreau** [comment]
TL;DR: some string keys (empty, but also those with delimiters) cannot be typed on the command line and thus cannot be removed either.

What we should do I think, is to add an optional `ptr` argument to pass the pointer to the key to remove.

**Comment by Darlelet** [comment]
It should be fixed now (https://github.com/haproxy/haproxy/commit/9b2717e7bb754c75816fb9272f3a2bcd6ea258cf is available starting with 3.0)

Should we close the issue?

**Comment by wtarreau** [comment]
Ah you're right, I forgot this one! Yes I think we should close it as fixed.

**Comment by Darlelet** [comment]
Thanks for confirming, now closing!
